source/common/upstream/upstream_impl.cc

```diff
@@ -34,6 +34,20 @@
 
 namespace Envoy {
 namespace Upstream {
+namespace {
+
+const Network::Address::InstanceConstSharedPtr
+getSourceAddress(const envoy::api::v2::Cluster& cluster,
+                 const Network::Address::InstanceConstSharedPtr source_address) {
+  // The source address from cluster config takes precedence.
+  if (cluster.has_upstream_bind_config() && cluster.upstream_bind_config().has_source_address()) {
```

Comment:
nit: Do we need `has_upstream_bind_config()` call? (Default object should do right thing).